### ISSUE TITLE
Release 3.2.0

### ISSUE THREAD

**Issue by zanchey** [fish-shell/fish-shell#6585]
Let's try and avoid another year between drinks! I think increasing the release cadence will actually shorten the amount of work for each release, so I'd be happy to try and do it in 2-3 months time.

Assuming the tree is pretty stable, let's aim for a release in April - I've got some time up my sleeve then, with a beta in the first week and a release 2-3 weeks later.

As always, deadlines subject to availability.

**Comment by ridiculousfish** [comment]
Sounds great! Also please do suggest anything that comes to your mind that can help make the releases more automated.

**Comment by zanchey** [comment]
Haha COVID-19 really blew this out of the water, didn't it!

**Comment by Pytal** [comment]
What's the status on this?

**Comment by faho** [comment]
The same as every release: It'll be out when it's out.

**Comment by zanchey** [comment]
I've got another project I'm finishing up then will make a start, but as always can promise no timeline.

**Comment by zanchey** [comment]
I'm making headway on the release notes, and I'm wondering about whether or not to do our usual two-week beta or just release it after cutting a separate branch. If I test the macOS builds and @andrew-schulman and @Mno-hime can test on Cygwin and OpenIndiana respectively that will cover most of the issues found in recent betas (looking back to 2.4.0).

I'm happy to take advice though!

**Comment by andrew-schulman** [comment]
I can test quickly in Cygwin.

**Comment by faho** [comment]
>I'm making headway on the release notes, and I'm wondering about whether or not to do our usual two-week beta or just release it after cutting a separate branch

Tbh we've not had great success with the betas - we only really got feedback after the actual release.

So I'd rather we just do a release and stay ready to do a quick follow-up .1 a week or month later, if that's possible?

**Comment by zanchey** [comment]
> So I'd rather we just do a release and stay ready to do a quick follow-up .1 a week or month later, if that's possible?

That would be my plan! Point releases are actually easier to do.

**Comment by faho** [comment]
Do we "freeze" master for a bit, i.e. stop adding potentially disruptive changes for a while so we can get something releasable?

**Comment by cole-h** [comment]
Alternatively, you could branch off to `fish-3.2` (or something), effectively freezing that branch but leaving master open for other changes.

**Comment by zanchey** [comment]
Yeah, I'll branch `Integration_3.2.0` when we're close and let it bake for a couple of weeks.

**Comment by TylerYep** [comment]
Are the tentative/in-progress release notes available to read anywhere?

**Comment by faho** [comment]
@TylerYep The changelog is just kept in the repo: https://github.com/fish-shell/fish-shell/blob/master/CHANGELOG.rst

**Comment by andreiborisov** [comment]
I have a plugin, that relies on a bugfix that landed in `main` but still not in production. Would love the new release🙌

**Comment by lra** [comment]
Can we expect a fish release soon? Fish completion is slow due to https://github.com/fish-shell/fish-shell/pull/7365.
This is super annoying and I'd like not to mess with my pristine install to speed up completion again.

**Comment by faho** [comment]
> This is super annoying and I'd like not to mess with my pristine install to speed up completion again.

Run this interactively:

```fish
function __fish_describe_command; end
funcsave __fish_describe_command
```

And it should be worked around - note that the fault lies with apple, who not only broke this in the first place, but then fixed it and broke it again after that.

The new release will happen when it happens.

**Comment by pvaibhav** [comment]
Hi all, just some feedback from a user's perspective: it's been several months since macOS 11 was released and about a year since this thread was created.

It would be nice if at least the tab-completion bug fix #7365 could be back ported into fish 3.1.x.

**Comment by rnhmjoj** [comment]
I have to agree with @pvaibhav: as a user, fish releases too infrequently to keep up with other command line programs. The completion are always out-of-date and bug takes months to be fixed.
For instance, I noticed and fixed this https://github.com/fish-shell/fish-shell/pull/6725 in march, but I have to apply a workaround locally nearly a year later.

I think bug fixes and completion updates should be backported to a stable branch that releases more frequently for a better user experience.

**Comment by miki725** [comment]
Just wanted to mentioned that granted that the a release will be welcome however for the impatient for the time being using the head version might be sufficient. For example for arch AUR has [`fish-git`](https://aur.archlinux.org/packages/fish-git/) or for mac `brew install fish --HEAD`. Other distros should have similar options to install a package from git.

**Comment by etcusrvar** [comment]
Running from latest in git is fine for standard distros. I recently added fish to Synology devices ([spksrc/#4366](https://github.com/SynoCommunity/spksrc/pull/4366), [package](https://synocommunity.com/package/fish)) which requires cross-compiling and manual installation. It would be nice to have a newer version there with some of the recent fish issues I've been involved in, plus ones others have mentioned. Sure, it's doable to cross-compile and manually install locally from git occasionally-- it just would be nicer to have an actual release to do it from for public consumption.

**Comment by faho** [comment]
Alright, there's really very little productive happening here.

We would like to have a new release, but it'll happen when it happens. Saying that you'd also like a new release isn't helpful.

>I think bug fixes and completion updates should be backported to a stable branch that releases more frequently for a better user experience.

We simply do not have enough people to handle multiple release tracks, and even if we did experience shows that it wouldn't be used much - Ubuntu is still on fish 3.1.**0** - they haven't even upgraded to the stable release yet!

I'm going to ask y'all to discuss other things elsewhere.

**Comment by zanchey** [comment]
OK, we're really close. I'll ask @fish-shell/contributors to hold off from pushing any non-essential changes to `Integration_3.2.0` 🥶 .

One thing I haven't done is read through every commit to make sure we haven't missed anything in the CHANGELOG. I believe that @faho and @krobelus have done so, but there's 900+ commits and I just don't have the time. I'm happy to release an errata document if needed.

The most useful thing anyone in this thread can do in the next 72 hours is to trial a development build, either by [following the instructions in the wiki](https://github.com/fish-shell/fish-shell/wiki/Development-builds) or [building this tarball](https://github.com/fish-shell/fish-shell/files/6017803/fish_3.1.2-2145-g0808e5094.tar.gz) (especially on Cygwin @andrew-schulman or Illumos @Mno-hime!).

Otherwise, with no show-stoppers arising, I'll aim for a release later in the week.

**Comment by andrew-schulman** [comment]
Cygwin build fails, with multiple errors. Sorry but unfortunately I'm not able to make much sense of them. Here's the build log:
[fish-3.1.2-2145-g0808e5094-1-compile.log](https://github.com/fish-shell/fish-shell/files/6017931/fish-3.1.2-2145-g0808e5094-1-compile.log)

**Comment by zanchey** [comment]
OK, most of that looks like `wchar_t` size problems. `O_ASYNC` isn't available on Cygwin, so that will need to be worked around as well.

**Comment by mqudsi** [comment]
The last time I checked, the codepath with `O_ASYNC` isn't actually used on Cygwin, but it's just required for the code to compile. 
I pushed a fix for that in 2cf5fd3d5dd20035d694f4f5116cf5bd48765cd1 but I didn't have a chance to see the char16_t stuff. I have Cygwin on my other machine; I can try to take a look tomorrow or the day after.

**Comment by zanchey** [comment]
Yep, that works!

**Comment by mqudsi** [comment]
I pushed an update (215df7eec63e058e535045396fbb725f0f29f4b0) with a fix for the `char16_t` warnings-converted-to-errors, which led me down the rabbit hole of trying to fix some obvious UTF-16 issues with surrogate pairs not being handled. Fish now builds correctly on Cygwin and I guess that matter can technically be considered closed.

That said: If anyone at any point ever cares, the attached patch file adds surrogate codepoint handling to `fish_key_reader`, but I failed to realize just how bad the status of UTF-16 support in fish was. The patch works but fish already bails and refuses to even execute `echo \U1F604` under Cygwin/UTF-16 with errors about invalid tokens. I imagine the entirety of the input handling would need to be overhauled to actually support UTF-16 with its surrogate pairs, as we assume a single symbol will never be more than one `wchar_t`  in length pretty much everywhere, but that's obviously not true on UTF-16 platforms (my concern is that it won't be true on UTF-32 platforms either, some day). The patch fixes that assumption when transcribing user input into shell-encoded `\uxxxx` output, but after observing the behavior I described I figured it wasn't worth adding it in if it's just a tiny bandaid over a gaping wound (which I only noticed because of the remaining compiler warnings).

```diff
diff --git a/src/fish_key_reader.cpp b/src/fish_key_reader.cpp
index 39c7bfe5e..9527bbd4c 100644
--- a/src/fish_key_reader.cpp
+++ b/src/fish_key_reader.cpp
@@ -135,7 +135,13 @@ static void ascii_printable_to_symbol(wchar_t *buf, int buf_len, wchar_t wc, boo
 
 /// Convert a wide-char to a symbol that can be used in our output. The use of a static buffer
 /// requires that the returned string be used before we are called again.
-static wchar_t *char_to_symbol(wchar_t wc, bool bind_friendly) {
+///
+/// In case of UTF-16 encoding, we can run into a surrogate pair where we need two wchar_t inputs
+/// before we can determine the output. Without real algebraic data types, a maybe_t::None should be
+/// understood as "need another byte", to be provided via the \p prev_surrogate parameter.
+/// In this case, it is OK to call char_to_symbol again without handling the contents of the buffer.
+static maybe_t<wchar_t *> char_to_symbol(wchar_t wc, bool bind_friendly,
+        __attribute__((unused)) const maybe_t<wchar_t> &prev_surrogate = none()) {
     static wchar_t buf[64];
 
     if (wc < L' ') {  // ASCII control character
@@ -146,11 +152,45 @@ static wchar_t *char_to_symbol(wchar_t wc, bool bind_friendly) {
         del_to_symbol(buf, sizeof(buf) / sizeof(*buf), wc, bind_friendly);
     } else if (wc < 0x80) {  // ASCII characters that are not control characters
         ascii_printable_to_symbol(buf, sizeof(buf) / sizeof(*buf), wc, bind_friendly);
-    } else if (wc <= 0xFFFF) {  // BMP Unicode chararacter
+    }
+// Conditional handling of BMP Unicode characters depends on the encoding. Assume width of wchar_t
+// corresponds to the encoding, i.e. WCHAR_T_BITS == 16 implies UTF-16 and WCHAR_T_BITS == 32
+// because there's no other sane way of handling the input.
+#if WCHAR_T_BITS == 16
+    else if (wc <= 0xD7FF || (wc >= 0xE000 && wc <= 0xFFFD)) {  // BMP Unicode chararacter
+        std::swprintf(buf, sizeof(buf) / sizeof(*buf), L"\\u%04X", wc);
+    } else if (prev_surrogate == L'\0') {
+        // We need another 16-bit character to be able to decode this
+        return none();
+    } else {
+        // We are expecting (prev, wc) to represent a surrogate pair, with one high surrogate and
+        // one low surrogate (in that order).
+        if (!(*prev_surrogate >= 0xD800 && *prev_surrogate <= 0xDBFF
+                    && wc >= 0xDC00 && wc <= 0xDFFF)) {
+            // Invalid combination of surrogate pairs
+            std::swprintf(buf, sizeof(buf) / sizeof(*buf), L"\\uFFFD");
+        } else {
+            // Decode the surrogate pair
+            char32_t codepoint = 0x10000;
+            codepoint += (*prev_surrogate & 0x03FF) << 10;
+            codepoint += (wc & 0x03FF);
+
+            if (codepoint <= 0xFFFF) {  // UTF-32 chararacter in BMP range
+                std::swprintf(buf, sizeof(buf) / sizeof(*buf), L"\\u%04X", codepoint);
+            } else {  // UTF-32 character in non-BMP range
+                std::swprintf(buf, sizeof(buf) / sizeof(*buf), L"\\U%06X", codepoint);
+            }
+        }
+    }
+#elif WCHAR_T_BITS == 32
+    else if (wc <= 0xFFFF) {  // BMP Unicode chararacter
         std::swprintf(buf, sizeof(buf) / sizeof(*buf), L"\\u%04X", wc);
     } else {  // Non-BMP Unicode chararacter
         std::swprintf(buf, sizeof(buf) / sizeof(*buf), L"\\U%06X", wc);
     }
+#else
+    static_assert(false, "Unsupported WCHAR_T size; unknown encoding!");
+#endif
 
     return buf;
 }
@@ -162,16 +202,41 @@ static void add_char_to_bind_command(wchar_t wc, std::vector<wchar_t> &bind_char
 static void output_bind_command(std::vector<wchar_t> &bind_chars) {
     if (!bind_chars.empty()) {
         std::fputws(L"bind ", stdout);
+        bool surrogate_pair = false;
+        maybe_t<wchar_t> prev_surrogate = none();
         for (auto bind_char : bind_chars) {
-            std::fputws(char_to_symbol(bind_char, true), stdout);
+            auto symbol = char_to_symbol(bind_char, true, prev_surrogate);
+            if (!symbol.has_value()) {
+                prev_surrogate = bind_char;
+                surrogate_pair = true;
+                continue;
+            }
+            surrogate_pair = false;
+            prev_surrogate = none();
+            std::fputws(symbol.value(), stdout);
+        }
+        if (surrogate_pair) {
+            // Unexpected end of input: print decode error symbol.
+            std::fputws(L"\uFFFD", stdout);
         }
         std::fputws(L" 'do something'\n", stdout);
         bind_chars.clear();
     }
 }
 
-static void output_info_about_char(wchar_t wc) {
-    std::fwprintf(stderr, L"hex: %4X  char: %ls\n", wc, char_to_symbol(wc, false));
+enum class char_output_info_t {
+    Ok,
+    NeedMore,
+};
+
+static char_output_info_t output_info_about_char(wchar_t wc,
+        const maybe_t<wchar_t> &prev_surrogate = none()) {
+    auto symbol = char_to_symbol(wc, false, prev_surrogate);
+    if (!symbol.has_value()) {
+        return char_output_info_t::NeedMore;
+    }
+    std::fwprintf(stderr, L"hex: %4X  char: %ls\n", wc, *symbol);
+    return char_output_info_t::Ok;
 }
 
 static bool output_matching_key_name(wchar_t wc) {
@@ -204,6 +269,7 @@ static void process_input(bool continuous_mode) {
     input_event_queue_t queue;
     std::vector<wchar_t> bind_chars;
 
+    maybe_t<wchar_t> prev_surrogate = none();
     std::fwprintf(stderr, L"Press a key:\n");
     for (;;) {
         char_event_t evt{0};
@@ -228,7 +294,11 @@ static void process_input(bool continuous_mode) {
         if (wc) {
             add_char_to_bind_command(wc, bind_chars);
         }
-        output_info_about_char(wc);
+        if (output_info_about_char(wc, prev_surrogate) == char_output_info_t::NeedMore) {
+            // Need another input character to decode this surrogate pair
+            continue;
+        }
+        prev_surrogate = none();
         if (output_matching_key_name(wc)) {
             output_bind_command(bind_chars);
         }
```

**Comment by faho** [comment]
>obviously not true on UTF-16 platforms (my concern is that it won't be true on UTF-32 platforms either, some day)

(I don't see any of those popping up. In fact I see even Windows moving to UTF-8 eventually)

Anyway, that's a matter for at the very least a different release.

**Comment by mqudsi** [comment]
> Anyway, that's a matter for at the very least a different release.

No argument from me and I certainly wasn't suggesting otherwise!

> (I don't see any of those popping up. In fact I see even Windows moving to UTF-8 eventually)

My concern is that there are only so many emoji you can fit into the remaining 32-bit Unicode range and they clearly show no sign of letting up.

**Comment by faho** [comment]
>My concern is that there are only so many emoji you can fit into the remaining 32-bit Unicode range and they clearly show no sign of letting up.

That's... not a problem of the encoding, and it's not something UTF-32 (which is 4 byte by nature) would help with.

I think the simplest solution would be to extend UTF-8 to 5 bytes (and eventually more), which gives us 128 times the space (given that one more bit would have to be used to mark the fifth bit).

Also Unicode 13 has 283,506 codepoints, leaving 830,606 unassigned. Given that it added 6000 in one year, that gives us 138 years until it's exhausted, and that is *if* it doesn't let up.

**Comment by mqudsi** [comment]
Sorry, I wasn't referring to Unicode but rather to the specific use of UTF-32. UTF-8 forces you to deal with "one unit may not contain a full glyph" but UTF-32 lets you get away with pretending it doesn't happen; which is awesome so long as it doesn't end up happening (cf UCS-2).

**Comment by andrew-schulman** [comment]
With 215df7e, 3.2.0 almost builds in Cygwin. There's still one error:

```
src/env_universal_common.cpp:1391:52: error: ‘O_ASYNC’ was not declared in this scope; did you mean ‘O_RSYNC’?
 1391 |         if (fcntl(pipe.fd(), F_SETFL, O_NONBLOCK | O_ASYNC) == -1) {
      |                                                    ^~~~~~~
      |                                                    O_RSYNC
```

**Comment by mqudsi** [comment]
@andrew-schulman I think you are missing 2cf5fd3d5dd20035d694f4f5116cf5bd48765cd1 ?

**Comment by mqudsi** [comment]
@zanchey @Mno-hime d95e99f2d4b0471a5b389a0325251f809dc8cbf2 fixes an issue with missing memset/memzero on illumos via OI, but userland is broken on termsize problems caused by curses vs ncurses differences. This is with a naive cmake build w/ gcc9, not sure if you have some build script that explicitly uses ncurses instead (but in all cases we should either fix our support for "real" curses or drop it).

**Comment by zanchey** [comment]
https://github.com/OpenIndiana/oi-userland/blob/0469b0aaecddb524ed550c5b955775a96232f624/components/shell/fish/Makefile#L48 appears to depend on ncurses, though I've no idea how that interacts with fish's slightly funky curses decision making.

**Comment by mqudsi** [comment]
I merged in some more fixes that should fix the build issues on Illumos. The ncurses dependency can probably be dropped from the package definition file.

**Comment by Mno-hime** [comment]
I fail to build on OpenIndiana using our [infrastructure](https://github.com/OpenIndiana/oi-userland/blob/0469b0aaecddb524ed550c5b955775a96232f624/components/shell/fish/Makefile) `Makefile`:
```
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/termsize.cpp: In function 'maybe_t<termsize_t> read_termsize_from_tty()':
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/termsize.cpp:17:20: error: variable 'read_termsize_from_tty()::winsize winsize' has initializer but incomplete type
     struct winsize winsize = {0, 0, 0, 0};
                    ^~~~~~~
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/termsize.cpp:18:30: error: 'TIOCGWINSZ' was not declared in this scope
     if (ioctl(STDOUT_FILENO, TIOCGWINSZ, &winsize) >= 0) {
                              ^~~~~~~~~~
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/termsize.cpp:18:30: note: suggested alternative: 'LIOCGETS'
     if (ioctl(STDOUT_FILENO, TIOCGWINSZ, &winsize) >= 0) {
                              ^~~~~~~~~~
                              LIOCGETS
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/termsize.cpp:28:59: error: no matching function for call to 'termsize_t::termsize_t(<brace-enclosed initializer list>)'
         result = termsize_t{winsize.ws_col, winsize.ws_row};
                                                           ^
In file included from /export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/termsize.cpp:3:0:
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/termsize.h:29:5: note: candidate: termsize_t::termsize_t(int, int)
     termsize_t(int w, int h) : width(w), height(h) {}
     ^~~~~~~~~~
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/termsize.h:29:5: note:   conversion of argument 1 would be ill-formed:
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/termsize.h:17:8: note: candidate: constexpr termsize_t::termsize_t(const termsize_t&)
 struct termsize_t {
        ^~~~~~~~~~
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/termsize.h:17:8: note:   candidate expects 1 argument, 2 provided
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/termsize.h:17:8: note: candidate: constexpr termsize_t::termsize_t(termsize_t&&)
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/termsize.h:17:8: note:   candidate expects 1 argument, 2 provided
```
I took the following patches on top of @zanchey's [tarball](https://github.com/fish-shell/fish-shell/issues/6585#issuecomment-782862498) 215df7eec63e058e535045396fbb725f0f29f4b0, 2cf5fd3d5dd20035d694f4f5116cf5bd48765cd1, and d95e99f2d4b0471a5b389a0325251f809dc8cbf2. Should I take more or some other tarball?

I also see the following warning (I may be hitting some illumos-specific path):
```
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/env_universal_common.cpp: In function 'bool get_mac_address(unsigned char*)':
/export/home/newman/oi-userland/components/shell/fish/fish-3.1.2/src/env_universal_common.cpp:1022:70: warning: unused parameter 'macaddr' [-Wunused-parameter]
 static bool get_mac_address(unsigned char macaddr[MAC_ADDRESS_MAX_LEN]) { return false; }
                                                                      ^
```
And this:
```
[ 59%] Building CXX object CMakeFiles/fishlib.dir/src/event.cpp.o
[ 60%] Building CXX object CMakeFiles/fishlib.dir/src/exec.cpp.o
At global scope:
cc1plus: warning: unrecognized command line option '-Wno-redundant-move'
```
The build system uses GCC 7.5.0.

Regarding ncurses on illumos: CMake found `/usr/lib/64/libcurses.so` which is part of illumos not ncurses (but I'll be more sure when fish actually builds :) ).

**Comment by mqudsi** [comment]
@Mno-hime you'll need ffa24eb3610dc2853c1a468b8aa2fd327cd0d3fa

**Comment by Mno-hime** [comment]
@mqudsi Yup, that worked. `ncurses` is gone from fish dependencies after package build. Running tests now.

**Comment by Mno-hime** [comment]
There were few test errors: [fish_oi.log](https://github.com/fish-shell/fish-shell/files/6037186/fish_oi.log). But fish tests were never clean on illumos.

**Comment by faho** [comment]
@Mno-hime: Please try `make test_fishscript` and `make test_interactive` (the latter will require pexpect now, not the old tcl expect).

Your failures are in the low-level tests (universal variable syncing and history), and I recall those never working as well, but it makes the build stop and so it never gets to the script and interactive tests.

**Comment by mqudsi** [comment]
The notifier/uvar failures can probably be fixed by using a different notifier strategy. I'm at unease with making the new sigio notifier the default for anything but Linux and maybe BSD as it seems to rely on too much "off the beaten path" voodoo.

**Comment by faho** [comment]
Alright, tests are passing on Github Actions (Ubuntu & macOS), sr.ht (Arch, Alpine, FreeBSD) and manually on Arch and NetBSD with b154ad6a5c33a96e5b39a09272f7e7a131efdb92 (full low-level, script and interactive on all).

**Comment by mqudsi** [comment]
I propose the following, at least for 3.2.0 but really no good reason not indefinitely given the portability and well-defined behavior of the named pipe notifier:

```diff
diff --git a/src/env_universal_common.cpp b/src/env_universal_common.cpp
index 36f69cc72..374adf98a 100644
--- a/src/env_universal_common.cpp
+++ b/src/env_universal_common.cpp
@@ -1582,7 +1582,7 @@ universal_notifier_t::notifier_strategy_t universal_notifier_t::resolve_default_
     return strategy_notifyd;
 #elif defined(__CYGWIN__)
     return strategy_shmem_polling;
-#elif defined(SIGIO)
+#elif defined(SIGIO) && (defined(__APPLE__) || defined(__BSD__) || defined(__linux__))
     // The SIGIO notifier does not yet work on WSL. See #7429
     if (is_windows_subsystem_for_linux()) {
         return strategy_named_pipe;
```

**Comment by Mno-hime** [comment]
I run with GNU tools in the `PATH` preceding illumos utilities (GNU ones support more options) produced two errors.
`test_fishscript`:
```
Testing file checks/jobs.fish ... Failure:

  The CHECKERR on line 42 wants:
    bg: invalid option -23

  which failed to match line stderr:1:
    ps: unknown output format: -o stat

  Context:
    ps: unknown output format: -o stat <= no check matches
    usage: ps [ -aAdefHlcjLPWyZ ] [ -o format ] [ -t termlist ] <= no check matches
    \t[ -u userlist ] [ -U userlist ] [ -G grouplist ] <= no check matches
    \t[ -p proclist ] [ -g pgrplist ] [ -s sidlist ] <= no check matches
    \t[ -z zonelist ] [-h lgrplist] <= no check matches
      \'format\' is one or more of: <= no check matches
    \taddr args c class comm ctid dmodel etime f fname gid group lgrp lwp <= no check matches
    \tlwpname nice nlwp opri osz pcpu pgid pid pmem ppid pri project projid <= no check matches
    \tpset psr rgid rgroup rss ruid ruser s sid stime taskid time tty uid <= no check matches
    \tuser vsz wchan zone zoneid <= no check matches
    ps: unknown output format: -o stat <= no check matches
    usage: ps [ -aAdefHlcjLPWyZ ] [ -o format ] [ -t termlist ] <= no check matches
    \t[ -u userlist ] [ -U userlist ] [ -G grouplist ] <= no check matches
    \t[ -p proclist ] [ -g pgrplist ] [ -s sidlist ] <= no check matches
    \t[ -z zonelist ] [-h lgrplist] <= no check matches
      \'format\' is one or more of: <= no check matches
    \taddr args c class comm ctid dmodel etime f fname gid group lgrp lwp <= no check matches
    \tlwpname nice nlwp opri osz pcpu pgid pid pmem ppid pri project projid <= no check matches
    \tpset psr rgid rgroup rss ruid ruser s sid stime taskid time tty uid <= no check matches
    \tuser vsz wchan zone zoneid <= no check matches
    bg: invalid option -23
    fg: No suitable job: 3
    bg: Could not find job \'3\'

  when running command:
    ../test/root/bin/fish checks/jobs.fish
```
```
Testing file checks/math.fish ... Failure:

  The CHECK on line 143 wants:
    2

  which failed to match line stdout:39:
    20

  additional output on stderr:26:49:
    '0x2'
       ^
    math: Error: Unknown function
    '2 x4'
        ^
    math: Error: Logical operations are not supported, use `test` instead
    '42 >= 1337'
        ^
    math: Error: Unknown function
    'bitand(0xFE, 1)'
             ^
    math: Error: Unknown function
    'bitor(0xFE, 1)'
            ^
    math: Error: Missing opening parenthesis
    'log 16'
          ^
    math: Error: Missing closing parenthesis
    'log(16'
          ^
    math: Error: Unknown function
    '12 + 0x50'
             ^
    math: 'notabase' is not a valid base value

  Context:
    [...] from line 87 (stdout:36):
    2
    -4.5
    -4
     <= nothing to match CHECK '2' on line 143
    20
    8
    0
     <= nothing to match CHECK '0' on line 154
     <= nothing to match CHECK '255' on line 156
    4
    0
    1
    69343999
    0x11
    0x20
     <= nothing to match CHECK '0x5c' on line 182
    067
    2

  when running command:
    ../test/root/bin/fish checks/math.fish
```
`test_interactive`:
```
Testing file pexpects/fg.py ... Failed to match pattern: SIGTSTP
fg.py:92: timeout from expect_str("SIGTSTP")


Escaped buffer:
\r\n^Z

When written to the tty, this looks like:
<-------

^Z
------->

Last 10 messages:
OUTPUT    2431.99 ms (Line 76): \rprompt 11>
 INPUT      +2.09 ms (Line 79): bind \cr 'fg >/dev/null 2>/dev/null'\n
OUTPUT     +17.35 ms (Line 80): \rprompt 12>
 INPUT      +1.53 ms (Line 81): $fish_test_helper print_stop_cont\n
 INPUT    +201.85 ms (Line 84): \x1a
OUTPUT      +4.25 ms (Line 85): SIGTSTP
OUTPUT      +9.44 ms (Line 85): \rprompt 13>
 INPUT    +103.27 ms (Line 87): \x12
OUTPUT      +3.87 ms (Line 88): SIGCONT
 INPUT      +2.62 ms (Line 91): \x1a

Trying pexpects/fg.py for a second time
```
```
Testing file pexpects/fg.py ... Failed to match pattern: SIGTSTP
fg.py:92: timeout from expect_str("SIGTSTP")


Escaped buffer:
\r\n^Z

When written to the tty, this looks like:
<-------

^Z
------->

Last 10 messages:
OUTPUT    2404.82 ms (Line 76): \rprompt 11>
 INPUT      +3.11 ms (Line 79): bind \cr 'fg >/dev/null 2>/dev/null'\n
OUTPUT     +13.74 ms (Line 80): \rprompt 12>
 INPUT      +2.36 ms (Line 81): $fish_test_helper print_stop_cont\n
 INPUT    +202.32 ms (Line 84): \x1a
OUTPUT      +3.68 ms (Line 85): SIGTSTP
OUTPUT      +9.45 ms (Line 85): \rprompt 13>
 INPUT    +102.36 ms (Line 87): \x12
OUTPUT      +4.03 ms (Line 88): SIGCONT
 INPUT      +2.28 ms (Line 91): \x1a
```
```
Testing file pexpects/fkr.py ... Failed to match pattern: Exiting\ at\ your\ request\.\\r\\n
fkr.py:51: timeout from expect_str("Exiting at your request.\r\n")


Escaped buffer:
(  0.127 ms)  hex:    4  char: \cD\r\nbind \cD\cD 'do something'\r\n

When written to the tty, this looks like:
<-------
(  0.127 ms)  hex:    4  char: \cD
bind \cD\cD 'do something'

------->

Last 10 messages:
OUTPUT      47.06 ms (Line 28): char: \cG  (or \a)\r\nbind \a 'do something'\r\n
 INPUT     +20.92 ms (Line 33): \x1bሴ
OUTPUT     +32.05 ms (Line 34): char: \u1234\r\nbind \e\u1234 'do something'\r\n
 INPUT     +22.66 ms (Line 38): \x00
OUTPUT      +4.30 ms (Line 39): char: \c@\r\nbind -k nul 'do something'\r\n
 INPUT      +2.65 ms (Line 42): \x03
 INPUT     +12.39 ms (Line 44): \x04
OUTPUT      +3.77 ms (Line 45): char: \cD\r\n
 INPUT     +52.50 ms (Line 49): \x04\x04
OUTPUT      +3.35 ms (Line 50): char: \cD\r\n

Trying pexpects/fkr.py for a second time
```
```
Testing file pexpects/fkr.py ... Failed to match pattern: Exiting\ at\ your\ request\.\\r\\n
fkr.py:51: timeout from expect_str("Exiting at your request.\r\n")


Escaped buffer:
(  0.030 ms)  hex:    4  char: \cD\r\nbind \cD\cD 'do something'\r\n

When written to the tty, this looks like:
<-------
(  0.030 ms)  hex:    4  char: \cD
bind \cD\cD 'do something'

------->

Last 10 messages:
OUTPUT      44.44 ms (Line 28): char: \cG  (or \a)\r\nbind \a 'do something'\r\n
 INPUT     +20.55 ms (Line 33): \x1bሴ
OUTPUT     +31.32 ms (Line 34): char: \u1234\r\nbind \e\u1234 'do something'\r\n
 INPUT     +20.52 ms (Line 38): \x00
OUTPUT      +1.70 ms (Line 39): char: \c@\r\nbind -k nul 'do something'\r\n
 INPUT      +0.53 ms (Line 42): \x03
 INPUT     +10.43 ms (Line 44): \x04
OUTPUT      +1.54 ms (Line 45): char: \cD\r\n
 INPUT     +50.54 ms (Line 49): \x04\x04
OUTPUT      +1.53 ms (Line 50): char: \cD\r\n
```
```
Testing file pexpects/terminal.py ... Failed to match pattern: term\-support\:\ Terminal\ has\ 0\ columns\,\ falling\ back\ to\ default\ width
terminal.py:20: timeout from expect_str("term-support: Terminal has 0 columns, falling back to default width")


Escaped buffer:
\r\n⏎                                                                              \r⏎ \r\rprompt 2>

When written to the tty, this looks like:
<-------

prompt 2>                                                                      
------->

Last 10 messages:
OUTPUT       0.00 ms (Line 16): \rprompt 1>
 INPUT       5.38 ms (Line 18): echo $COLUMNS $LINES\n
OUTPUT      +3.73 ms (Line 19): 80 24

Trying pexpects/terminal.py for a second time
```
```
Testing file pexpects/terminal.py ... Failed to match pattern: term\-support\:\ Terminal\ has\ 0\ columns\,\ falling\ back\ to\ default\ width
terminal.py:20: timeout from expect_str("term-support: Terminal has 0 columns, falling back to default width")


Escaped buffer:
\r\n⏎                                                                              \r⏎ \r\rprompt 2>

When written to the tty, this looks like:
<-------

prompt 2>                                                                      
------->

Last 10 messages:
OUTPUT       0.00 ms (Line 16): \rprompt 1>
 INPUT       5.23 ms (Line 18): echo $COLUMNS $LINES\n
OUTPUT      +2.71 ms (Line 19): 80 24
```

And btw two `fish_key_reader` processes keep running, each taking 100 % of one core. `truss` says they do this:
```
read(0, 0xFFFFFC7FFFDEE230, 1)			= 0
pollsys(0xFFFFFC7FFFDF01E0, 1, 0xFFFFFC7FFFDF02E0, 0x00000000) = 1
pollsys(0xFFFFFC7FFFDEE010, 2, 0x00000000, 0x00000000) = 1
```

**Comment by faho** [comment]
>I run with GNU tools in the PATH preceding illumos utilities (GNU ones support more options) produced two errors.

Well, your ps isn't any I've ever seen.

Does it support `ps -o state`, perchance?

>    '0x2'
       ^
    math: Error: Unknown function

Your wcstod isn't C++11 conforming IIUC - it requires `0xdeadbeef` hex numbers support.

>Testing file pexpects/fg.py ... Failed to match pattern: SIGTSTP

Apparently ctrl-z doesn't send sigstp here. Either it's a pexpect issue (not impossible), or your terminal is set up differently by default. What does `stty` say?

>Testing file pexpects/fkr.py ... Failed to match pattern: Exiting\ at\ your\ request\.\\r\\n

Same but ctrl-d/eof. This is where the fish_key_reader processes are coming from.

I'd check what ctrl-d does for you, because it seems they're not getting the signal they usually do.

>Testing file pexpects/terminal.py ... Failed to match pattern: term\-support\:\ Terminal\ has\ 0\ columns\,\ falling\ back\ to\ default\ width

Pexpect apparently doesn't manage to set the pty dimensions to 0/0. Should be uncritical.

**Comment by mqudsi** [comment]
Is there a reason we use wcstod w/ all our patches and not wstringstream? I detest stringstreams (who doesn't) but I imagine the more limited choice of C++ libraries would result in more consistent behavior, and I don't think we use literal to numeric conversions in any hot paths.. (obviously not for 3.2.0).

**Comment by faho** [comment]
>Is there a reason we use wcstod w/ all our patches and not wstringstream?

1. We don't currently use streams *anywhere*, AFAICT.
2. They are sloooooow.
3. The locale stuff is super awkward.
4. There's little guarantee they wouldn't be broken as well - given that they don't have a working wcstod to fall back on.
5. This would have been more useful *before* we added all the stuff to wcstod, because now it works basically everywhere, and where it doesn't I'm comfortable blaming the platform and letting them fix it. This isn't a case of the platform interpreting the spec differently, this is a case of the spec being partially implemented (and it's a part that our users there can just do without, just like openbsd won't obey LC_NUMERIC)

So this would *not* be worth it.

>I don't think we use literal to numeric conversions in any hot paths

`test`, `math` are frequently called in loops and may invoke this.

**Comment by dany74q** [comment]
Congrats on the release - impressive work !

**Comment by faho** [comment]
Just for posterity:

>Your wcstod isn't C++11 conforming IIUC - it requires 0xdeadbeef hex numbers support.

@Mno-hime I've confirmed this is the case - [C11 says](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1548.pdf) in §7.22.1.3:

> The expected form of the subject sequence is an optional plus or minus sign, then one of the following
[...]
> a 0x or 0X, then a nonempty sequence of hexadecimal digits optionally containing a decimal-point character, then an optional binary exponent part as defined in 6.4.4.2;

It's a conformance bug in your libc. (and C++11 is based on C11, so since we require C++11 support this is technically not fulfilling fish's requirements)

**Comment by xfix** [comment]
C++11 is based on C99, when C++11 was published there was no C11 standard.

**Comment by faho** [comment]
Ah, okay, it seems I looked at a newer version of C++. No matter, it's also in [C99](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) - §7.24.4.1.1. Also I'd be happy to add "also C11" to our requirements, given that it's also 10 years old.

**Comment by andrew-schulman** [comment]
3.20. builds and runs fine on Cygwin.

I looked at the Release Notes. Y'all have been busy. Good work, team.

I noticed that there are new files in our vendor_*.d directories, named .keep-fish-3.2.0-1. Do I actually need to keep those?

**Comment by andrew-schulman** [comment]
> I noticed that there are new files in our vendor_*.d directories, named .keep-fish-3.2.0-1. Do I actually need to keep those?

Nvm. Cygwin packaging thing.
